### PR TITLE
Adding validation capabilities to the napalm-base

### PR DESCRIPTION
**Pull request by GGabriele**
Refers to #83

Here I'm adding validation capabilities to napalm-base. This means that we'll be able to validate getters method results and running/candidate configuration as well.

Sample `validate.yml`

```

---

get_interfaces:
  Ethernet2/5:
    is_enabled: true
    is_up: false

  loopback103:
    is_enabled: false
    is_up: false

  Vlan100:
    is_enabled: false
    is_up: false

get_facts:
  os_version: 7.0(3)I2(2d)
  interface_list:
    - Vlan5
    - Vlan100
  hostname: n9k2

get_lldp_neighbors:
  mgmt0:
    - hostname: PERIMETER
      port: Fa1/0/3
    - hostname: nxos-spine2
      port: nxos-spine2

get_bgp_neighbors:
  default:
    router_id: 192.0.2.2
    peers:
      192.0.2.3:
        is_enabled: false
      192.0.2.2:
        is_enabled: false
        address_family:
          ipv4:
            sent_prefixes: 5
          ipv6:
            sent_prefixes: 2

get_arp_table:
  - interface: Ethernet2/1
    ip: 192.0.2.2

get_interfaces_ip:
  Ethernet2/1:
    ipv4:
      192.0.2.1:
        prefix_length: 30

running:
  - username gabriele password gabriele
  - interface Ethernet2/1
```

Sample implementation into NXOS driver:

```
    def validate(self, getters=None, config=None, validation_file='validate.yml'):
        errors = napalm_base.validate.validate(self, getters=getters, config=config, validation_file=validation_file)
        return errors
```

Sample workflow:

```
>>>import json
>>>from napalm import get_network_driver
>>>driver = get_network_driver('nxos')
>>>device = driver('nxos', 'ntc', 'ntc123')
>>>device.open()
>>>
>>>getters = ['get_facts', 'get_lldp_neighbors', 'get_bgp_neighbors', 'get_interfaces', 'get_interfaces_ip']
>>>validation = device.validate(getters=getters)
>>>
>>>print json.dumps(validation, indent=4)
```

Sample results:

```
{
    "get_facts": {
        "os_version": [
            "Expected key but not found."
        ], 
        "interface_list": [
            "Expected but not found: Vlan5", 
            "Expected but not found: Vlan100"
        ], 
        "hostname": [
            "Expected 'n9k2', found 'nxos-spine1' instead"
        ]
    }, 
    "get_interfaces": {
        "Ethernet2/5": [
            "Expected 'is_up: False', found 'is_up: True' instead"
        ], 
        "Vlan100": [
            "Expected key but not found."
        ], 
        "loopback103": [
            "Expected key but not found."
        ]
    }, 
    "get_interfaces_ip": {
        "Ethernet2/1": [
            {
                "ipv4": {
                    "192.0.2.1": [
                        "Expected 'prefix_length: 30', found 'prefix_length: 24' instead"
                    ]
                }
            }
        ]
    }, 
    "get_lldp_neighbors": {
        "mgmt0": [
            "Expected key but not found."
        ]
    }, 
    "get_bgp_neighbors": {
        "default": [
            {
                "router_id": "Expected '192.0.2.2', found '192.0.2.1' instead", 
                "peers": {
                    "192.0.2.3": "Expected key but not found.", 
                    "192.0.2.2": {
                        "is_enabled": "Expected 'False', found 'True' instead", 
                        "address_family": {
                            "ipv4": [
                                "Expected 'sent_prefixes: 5', found 'sent_prefixes: -1' instead"
                            ], 
                            "ipv6": "Expected key but not found."
                        }
                    }
                }
            }
        ]
    }
}
```

Later or tomorrow I'll add a sample workflow for configuration as well.
Any suggestion is welcome :)

### CODE REVIEW

**Comment by dbarrosop** [comment]
This PR needs some love in the form of documentation and tests ;)

**Comment by GGabriele** [comment]
Indeed :) I pushed it as a starting point

**Comment by GGabriele** [comment]
@dbarrosop can you take a look at the docstring?
Also, how do you think we should test this? Mock?

**Comment by dbarrosop** [comment]
Yeah, I am sure this one can be heavily tested with mocks. Standby on this one as we are thinking about creating a collection of utilities and this one might fit right in there rather than having it on napalm-base. You can still work on the tests in the meantime.

**Comment by dbarrosop** [comment]
I think the dosctring is good although I would like to see more examples, maybe a tutorial on read the docs would be better. @GGabriele proceed with the mocks and we will move this one is completed https://github.com/napalm-automation/napalm-utils

**Comment by GGabriele** [comment]
Okay cool, I'm (slowly :\ ) working on it.

**Comment by GGabriele** [comment]
I noticed we're getting closer to a new testing framework for mocks. Should I wait for it to be merged? 
Right now I have something like this in `base.py`. Not sure if having a test function for each getter is the best solution...what do you think?

```
class TestValidate:

    def _get_expected(self, getter):
        with open('{0}/{1}.txt'.format(self.vendor, getter), 'r') as f:
            return json.loads(f.read().replace("'", '"'))

    def test_validate_get_facts(self):
        try:
            self.device.get_facts()
            result = self.device.validate(getters=['get_facts'])
        except NotImplementedError:
            raise SkipTest()
        expected = self._get_expected('get_facts')
        self.assertTrue(result == expected)
```

**Comment by dbarrosop** [comment]
let me review this next week, I still think it would be idea to move this code to `napalm-tools` or something like that but will advise you next week : )

File: napalm_base/validate.py [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+                    else:
+                        not_matched.extend(not_matched_result)
+
+                if not_matched:
+                    not_matched_getter[key] = not_matched
+
+            if not_matched_getter:
+                errors[getter] = not_matched_getter
+
+    elif config:
```
Comment (dbarrosop) [inline napalm_base/validate.py]:
```
if getters:
...
elif config:
...
```

does that mean that either you validate getters or configuration?

**Comment by dbarrosop** [comment]
Sorry for having taken so long to look at this. I think this is super cool. I would like a couple of things and I can help with some:

1. Add it as a cli tool as well: https://github.com/napalm-automation/napalm-base/tree/develop/napalm_base/clitools
2. Write an ansible module. I think an ansible module to write assertions is gonna be great.
3. Add tests. This is not gonna be using the testing framework as it's specific to `napalm-base` only

**Comment by dbarrosop** [comment]
I have created a branch `validate` and change the PR so it merges there. I am gonna be merging it and I will see if I can give you a hand with this.

**Pull request merged**